src/extensions/averagewaveforms2/AverageWaveformsView/waveformLayer.ts

```diff
@@ -0,0 +1,53 @@
+import { funcToTransform } from '../../CanvasWidget';
+import { CanvasPainter } from '../../CanvasWidget/CanvasPainter';
+import { CanvasWidgetLayer } from "../../CanvasWidget/CanvasWidgetLayer";
+import { LayerProps } from './AverageWaveformWidget';
+import setupElectrodes, { ElectrodeBox } from './setupElectrodes';
+
+export type WaveformColors = {
+    base: string
+}
+const defaultWaveformColors: WaveformColors = {
+    base: 'black'
+}
+type LayerState = {
+    electrodeBoxes: ElectrodeBox[]
+}
+const initialLayerState = {
+    electrodeBoxes: []
+}
+
+export const createWaveformLayer = () => {
+    const onPaint = (painter: CanvasPainter, props: LayerProps, state: LayerState) => {
+        const { waveform } = props
+        const opts = props.waveformOpts
+        const colors = opts.colors || defaultWaveformColors
+        const maxAbs = Math.max(...waveform.map(w => Math.max(...w.map(x => Math.abs(x)))))
+        painter.wipe()
+        const yScaleFactor = 1 / maxAbs
+        for (let i = 0; i < state.electrodeBoxes.length; i++) {
+            const e = state.electrodeBoxes[i]
+            const painter2 = painter.transform(e.transform).transform(funcToTransform(p => {
```

Comment:
Create a transformed painter based on the transform matrix that is now stored with every electrode box.

Reply:
Note that there are 2 transforms. One to map into the electrode box, the other to map from waveform coordinates to [xmin: 0, xmax: 1, ymin: 0, ymax: 1] box coordingate

Reply:
Sure, do you mean comments within the code?

Reply:
Right, that's what I was thinking.

Reply:
Let's open an issue that puts this transform (that maps a rect to a rect) into the geometry module for canvas widget.